packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -357,15 +408,37 @@ export const changeSectionValueFormat = <TDate>(
   valueStr: string,
   currentFormat: string,
   newFormat: string,
-) => utils.formatByString(utils.parse(valueStr, currentFormat)!, newFormat);
+) => {
+  if (process.env.NODE_ENV !== 'production') {
+    if (getDateSectionConfigFromFormatToken(utils, currentFormat).dateSectionName === 'weekDay') {
+      throw new Error("changeSectionValueFormat doesn't support week day formats");
+    }
+  }
+
+  return utils.formatByString(utils.parse(valueStr, currentFormat)!, newFormat);
+};
 
 export const doesSectionHaveTrailingZeros = <TDate>(
   utils: MuiPickersAdapter<TDate>,
   contentType: 'digit' | 'letter',
-  formatValue: string,
-) =>
-  contentType === 'digit' &&
-  changeSectionValueFormat(utils, '1', formatValue, formatValue).length > 1;
+  dateSectionName: MuiDateSectionName,
+  format: string,
+) => {
+  if (contentType !== 'digit') {
+    return false;
+  }
+
+  if (dateSectionName === 'weekDay') {
+    return utils.formatByString(utils.startOfWeek(utils.date()!), format).length > 1;
+  }
+
+  // We can use `changeSectionValueFormat`, because  `utils.parse('1', 'YYYY')` returns `1971` instead of `1`.
+  if (dateSectionName === 'year') {
```

Comment:
Fixes #6590

I can extract it to a follow up PR but the fix seems simple enough to apply it here